src/promptflow/promptflow/batch/_base_executor_proxy.py

```diff
@@ -29,31 +35,88 @@
 
 class AbstractExecutorProxy:
     @classmethod
-    def get_tool_metadata(cls, flow_file: Path, working_dir: Optional[Path] = None) -> dict:
-        """Generate tool metadata file for the specified flow."""
-        return cls._get_tool_metadata(flow_file, working_dir or flow_file.parent)
+    def dump_metadata(cls, flow_file: Path, working_dir: Path) -> NoReturn:
+        """Generate metadata for a specific flow."""
+        cls.generate_flow_tools_json(flow_file, working_dir, dump=True)
+        cls.generate_flow_json(flow_file, working_dir, dump=True)
 
-    def _get_flow_meta(self) -> dict:
-        """Get the flow metadata from"""
+    @classmethod
+    def generate_flow_tools_json(
+        cls,
+        flow_file: Path,
+        working_dir: Path,
+        dump: bool = True,
+        timeout: int = FLOW_TOOLS_JSON_GEN_TIMEOUT,
+        load_in_subprocess: bool = True,
+    ) -> dict:
+        """Generate flow.tools.json for the specified flow."""
+        from promptflow import load_flow
+        from promptflow._sdk.entities._eager_flow import FlexFlow
+
+        flow = load_flow(flow_file)
+        if isinstance(flow, FlexFlow):
+            return {}
```

Comment:
Please refine here and do not depend on `FlexFlow`.

Reply:
updated

Reply:
Can pass `flow file` to `is_flex_flow`?

Reply:
here the problem is that, `flow` as a dictionary is potentially referred in context. It looks not a good practice to read files frequently unless necessary. Actually, I think the ideal parameter to be passed in to ExecutorProxy should be a `dict` object instead of a file path, so that we can load flow for only once and prevent potential impact of external file changes.

Reply:
Or let us see sample code for current batch engine logic:

Solution 1, flow_file as parameter, we will read the file for twice
```python
        ...
        with open(flow_file, "r", encoding="utf-8") as fin:
            flow_dag = load_yaml(fin)
        language = flow_dag.get(LANGUAGE_KEY, FlowLanguage.Python)
        return is_flex_flow(flow_file), language
```

Solution 2, 1 more function accept flow_file as input, then `is_flex_flow_dict` will have nearly the same functionality comparing to `is_flex_flow`:
```python
        ...
        with open(flow_file, "r", encoding="utf-8") as fin:
            flow_dag = load_yaml(fin)
        language = flow_dag.get(LANGUAGE_KEY, FlowLanguage.Python)
        return is_flex_flow_dict(flow_dag), language
```

Solution 3, make the function accept a union type of flow_dag. The trick part is, `load_yaml(xxx)` may also return a string in some cases, so we can't verify if the input is expected to be a path:
```python
def is_flex_flow(flow: Union[str, Path, dict]):
    ...
```

Solution 4, make the function accept 2 inputs. The problem is that the 2 parameters must be both optional and we will need extra logic to judge if none of them are provided:
```python
def is_flex_flow(*, file_path: Path = None, yaml_dict: dict = None):
    if file_path is None and yaml_dict is None:
        raise UserErrorException("xxx")
    ...
```

Hmm... I think only solution 4 is acceptable, but kind of complicated for such an util function

Reply:
updated